pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,907 @@
+                ..
+            }) = &stream.pending_request
+            {
+                now = min(now, *deadline);
+            }
+
+            // Dont perform payment
+            if now == last_time_updated {
+                return Ok(0u32.into());
+            }
+
+            // If deposit is zero the stream is fully drained and there is nothing to transfer.
+            if stream.deposit.is_zero() {
+                return Ok(0u32.into());
+            }
+
+            let delta = now
+                .checked_sub(&last_time_updated)
+                .ok_or(Error::<T>::TimeMustBeIncreasing)?;
+
+            // We compute the amount due to the target according to the rate, which may be
+            // lowered if the stream deposit is lower.
+            // Saturating is fine as it'll be clamped to the source deposit. It is also safer as
+            // considering it an error can make a stream un-updatable if too much time has passed
+            // without updates.
+            let mut payment = delta.saturating_mul(stream.config.rate);
+
+            // We compute the new amount of locked funds. If it underflows it
+            // means that there is more to pay that what is left, in which case
+            // we pay all that is left.
+            let (new_locked, drained) = match stream.deposit.checked_sub(&payment) {
+                Some(v) if v.is_zero() => (v, true),
+                Some(v) => (v, false),
+                None => {
+                    payment = stream.deposit;
+                    (Zero::zero(), true)
+                }
+            };
+
+            if payment.is_zero() {
+                return Ok(0u32.into());
+            }
+
+            // Transfer from the source to target.
+            T::Assets::transfer_deposit(
+                &stream.config.asset_id,
+                &stream.source,
+                &stream.target,
+                payment,
+            )?;
+
+            // Update stream info.
+            stream.deposit = new_locked;
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamPayment {
+                stream_id,
+                source: stream.source.clone(),
+                target: stream.target.clone(),
+                amount: payment,
+                drained,
+            });
+
+            Ok(payment)
+        }
+
+        fn apply_deposit_change(
+            stream: &mut StreamOf<T>,
+            change: DepositChange<T::Balance>,
+        ) -> DispatchResultWithPostInfo {
+            match change {
+                DepositChange::Absolute(amount) => {
+                    if let Some(increase) = amount.checked_sub(&stream.deposit) {
+                        T::Assets::increase_deposit(
+                            &stream.config.asset_id,
+                            &stream.source,
+                            increase,
+                        )?;
+                    } else if let Some(decrease) = stream.deposit.checked_sub(&amount) {
```

Comment:
Could you rewrite this in a way that returns an error if both calls to `checked_sub` fail? I know that's impossible but it makes me a bit nervous. For example:

```suggestion
                    } else {
                        let decrease = stream.deposit.checked_sub(&amount).ok_or(Error)?;
```

Reply:
Well if the requested amount equals the current amount, there is no issue and no deposit change should be made. It may occur if the request is made, then an immediate request top change the deposit is made.

Reply:
There is no issue here, just that I don't like seeing `if / else if` without a final `else`. Feel free to ignore me